pkg/registry/certificates/strategy.go

```diff
@@ -0,0 +1,207 @@
+)
+
+// csrStrategy implements behavior for CSRs
+type csrStrategy struct {
+	runtime.ObjectTyper
+	api.NameGenerator
+}
+
+// csrStrategy is the default logic that applies when creating and updating
+// CSR objects.
+var Strategy = csrStrategy{api.Scheme, api.SimpleNameGenerator}
+
+// NamespaceScoped is true for CSRs.
+func (csrStrategy) NamespaceScoped() bool {
+	return true
+}
+
+// AllowCreateOnUpdate is false for CSRs.
+func (csrStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+// PrepareForCreate clears fields that are not allowed to be set by end users
+// on creation. Users cannot create any derived information, but we expect
+// information about the requesting user to be injected by the registry
+// interface. Clear everything else.
+// TODO: check these ordering assumptions. better way to inject user info?
+func (csrStrategy) PrepareForCreate(obj runtime.Object) {
+	csr := obj.(*certificates.CertificateSigningRequest)
+	csr.Spec.Fingerprint = ""
+	csr.Spec.Subject = certificates.Subject{}
+	csr.Spec.Hostnames = []string{}
+	csr.Spec.IPAddresses = []string{}
+
+	// Be explicit that users cannot create pre-approved certificate requests.
+	csr.Status = certificates.CertificateSigningRequestStatus{}
+	csr.Status.Conditions = []certificates.CertificateSigningRequestCondition{}
+}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users
+// on update. Certificate requests are immutable after creation except via subresources.
+func (csrStrategy) PrepareForUpdate(obj, old runtime.Object) {
+	newCSR := obj.(*certificates.CertificateSigningRequest)
+	oldCSR := old.(*certificates.CertificateSigningRequest)
+
+	newCSR.Spec = oldCSR.Spec
+	newCSR.Status = oldCSR.Status
+}
+
+// Validate validates a new CSR. Validation must check for a correct signature.
+func (csrStrategy) Validate(ctx api.Context, obj runtime.Object) field.ErrorList {
+	csr := obj.(*certificates.CertificateSigningRequest)
+	return validation.ValidateCertificateSigningRequest(csr)
+}
+
+// Canonicalize normalizes the object after validation (which includes a
+// signature check). This ordering makes it an ideal place to extract some
+// trusted CSR information.
+func (csrStrategy) Canonicalize(obj runtime.Object) {
+	csr := obj.(*certificates.CertificateSigningRequest)
+
+	// TODO: assuming this all works because validation just did it
+	pemBytes, _ := base64.StdEncoding.DecodeString(csr.Spec.Request)
+	block, _ := pem.Decode(pemBytes)
+	request, _ := x509.ParseCertificateRequest(block.Bytes)
+
+	// internal.Subject is just a marshalling wrapper around pkix.Name
+	csr.Spec.Subject = certificates.NewInternalSubject(request.Subject)
+	csr.Spec.Hostnames = request.DNSNames
+	for _, ip := range request.IPAddresses {
+		csr.Spec.IPAddresses = append(csr.Spec.IPAddresses, ip.String())
+	}
+	fingerprint, err := calculateFingerprint(request)
+	if err != nil {
+		glog.Errorf("failed to calculate fingerprint of CSR %s: %v", csr.Name, err)
```

Comment:
It's certainly weird, but a lack of any of this derived information doesn't necessarily stop the admin from making a decision.